source/fuzz/transformation_flatten_conditional_branch.cpp

```diff
@@ -0,0 +1,604 @@
+    return false;
+  }
+
+  // The header block must end with an OpBranchConditional instruction.
+  if (header_block->terminator()->opcode() != SpvOpBranchConditional) {
+    return false;
+  }
+
+  // Use a set to keep track of the instructions that require fresh ids.
+  std::set<opt::Instruction*> instructions_that_need_ids;
+
+  // Check that, if there are enough ids, the conditional can be flattened.
+  if (!ConditionalCanBeFlattened(ir_context, header_block,
+                                 &instructions_that_need_ids)) {
+    return false;
+  }
+
+  // Get the mapping from instructions to the fresh ids available for them.
+  auto instructions_to_fresh_ids = GetInstructionsToFreshIdsMapping(ir_context);
+
+  {
+    // Check that all the ids given are fresh and distinct.
+
+    std::set<uint32_t> used_fresh_ids;
+
+    // Check the overflow ids.
+    for (uint32_t id : message_.overflow_id()) {
+      if (!CheckIdIsFreshAndNotUsedByThisTransformation(id, ir_context,
+                                                        &used_fresh_ids)) {
+        return false;
+      }
+    }
+
+    // Check the ids in the map.
+    for (const auto& pair : instructions_to_fresh_ids) {
+      for (uint32_t id : pair.second) {
+        if (!CheckIdIsFreshAndNotUsedByThisTransformation(id, ir_context,
+                                                          &used_fresh_ids)) {
+          return false;
+        }
+      }
+    }
+  }
+
+  // Keep track of the number of overflow ids still available in the overflow
+  // pool, as we go through the instructions.
+  int remaining_overflow_ids = message_.overflow_id_size();
+
+  for (auto instruction : instructions_that_need_ids) {
+    // The number of ids needed depends on the id of the instruction.
+    uint32_t ids_needed_by_this_instruction =
+        NumOfFreshIdsNeededByInstruction(ir_context, *instruction);
+    if (instructions_to_fresh_ids.count(instruction) != 0) {
+      // If there is a mapping from this instruction to a list of fresh
+      // ids, the list must have enough ids.
+
+      if (instructions_to_fresh_ids[instruction].size() <
+          ids_needed_by_this_instruction) {
+        return false;
+      }
+    } else {
+      // If there is no mapping, we need to rely on the pool of
+      // overflow ids, where there must be enough remaining ids.
+
+      remaining_overflow_ids -= ids_needed_by_this_instruction;
+
+      if (remaining_overflow_ids < 0) {
+        return false;
+      }
+    }
+  }
+
+  // All checks were passed.
+  return true;
+}
+
+void TransformationFlattenConditionalBranch::Apply(
+    opt::IRContext* ir_context,
+    TransformationContext* transformation_context) const {
+  uint32_t header_block_id = message_.header_block_id();
+  auto header_block = ir_context->cfg()->block(header_block_id);
+
+  // Find the first block where flow converges (it is not necessarily the merge
+  // block).
+  uint32_t convergence_block_id = header_block->MergeBlockId();
```

Comment:
I have commented on (1) - this is necessary because we cannot duplicate common blocks, and changed the algorithm so that the first point where flow diverges is chosen, as I have found out that it is possible to have an inner construct after flow has converged but before the merge block. So I don't think that I need to say anything about (2) because that case is now avoided